### **Improving Stability of Halide Perovskite Solar cells**

#### Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

#### Load Data

In [4]:
data = pd.read_excel('Data/Perovskite_database_content_all_data.xlsx')
data.head()

Ref_ID  Ref_ID_temp Ref_name_of_person_entering_the_data  \
0       1            1                       Adam Hultqvist   
1       2            2                       Adam Hultqvist   
2       3            3                       Adam Hultqvist   
3       4            4                       Adam Hultqvist   
4       5            5                       Adam Hultqvist   

   Ref_data_entered_by_author     Ref_DOI_number Ref_lead_author  \
0                       False  10.1021/jp5126624    Sabba et al.   
1                       False  10.1021/jp5126624    Sabba et al.   
2                       False  10.1021/jp5126624    Sabba et al.   
3                       False  10.1021/jp5126624    Sabba et al.   
4                       False  10.1021/jp5126624    Sabba et al.   

  Ref_publication_date                          Ref_journal  \
0           2015-01-06  The Journal of Physical Chemistry C   
1           2015-01-06  The Journal of Physical Chemistry C   
2           2015-01-06  The Journal of Physical Chemistry C   
3           2015-01-06  The Journal of Physical Chemistry C   
4           2015-01-06  The Journal of Physical Chemistry C   

   Ref_part_of_initial_dataset       Ref_original_filename_data_upload  ...  \
0                         True  Historic dataset on 2020 11 22_v7.xlsx  ...   
1                         True  Historic dataset on 2020 11 22_v7.xlsx  ...   
2                         True  Historic dataset on 2020 11 22_v7.xlsx  ...   
3                         True  Historic dataset on 2020 11 22_v7.xlsx  ...   
4                         True  Historic dataset on 2020 11 22_v7.xlsx  ...   

  Outdoor_PCE_Tse80 Outdoor_PCE_after_1000_h Outdoor_power_generated  \
0               NaN                      NaN                     NaN   
1               NaN                      NaN                     NaN   
2               NaN                      NaN                     NaN   
3               NaN                      NaN                     NaN   
4               NaN                      NaN                     NaN   

   Outdoor_link_raw_data_for_outdoor_trace  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                      NaN   

   Outdoor_detaild_weather_data_available  Outdoor_link_detailed_weather_data  \
0                                     0.0                                 NaN   
1                                     0.0                                 NaN   
2                                     0.0                                 NaN   
3                                     0.0                                 NaN   
4                                     0.0                                 NaN   

  Outdoor_spectral_data_available  Outdoor_link_spectral_data  \
0                             0.0                         NaN   
1                             0.0                         NaN   
2                             0.0                         NaN   
3                             0.0                         NaN   
4                             0.0                         NaN   

   Outdoor_irradiance_measured  Outdoor_link_irradiance_data  
0                          0.0                           NaN  
1                          0.0                           NaN  
2                          0.0                           NaN  
3                          0.0                           NaN  
4                          0.0                           NaN  

[5 rows x 410 columns]

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43025 entries, 0 to 43024
Columns: 410 entries, Ref_ID to Outdoor_link_irradiance_data
dtypes: bool(6), datetime64[ns](1), float64(150), int64(3), object(250)
memory usage: 132.9+ MB


A Total of 43025 data points with 410 features.

Keep only the Perovskite columns.

In [5]:
starting_letter = 'P'
selected_columns = [col for col in data.columns if col.startswith(starting_letter)]
df_filtered = data[selected_columns]

In [7]:
df_filtered.head()

Perovskite_single_crystal  Perovskite_dimension_0D  \
0                        0.0                      0.0   
1                        0.0                      0.0   
2                        0.0                      0.0   
3                        0.0                      0.0   
4                        0.0                      0.0   

   Perovskite_dimension_2D  Perovskite_dimension_2D3D_mixture  \
0                      0.0                                0.0   
1                      0.0                                0.0   
2                      0.0                                0.0   
3                      0.0                                0.0   
4                      0.0                                0.0   

   Perovskite_dimension_3D  Perovskite_dimension_3D_with_2D_capping_layer  \
0                     True                                            0.0   
1                     True                                            0.0   
2                     True                                            0.0   
3                     True                                            0.0   
4                     True                                            0.0   

  Perovskite_dimension_list_of_layers  \
0                                   3   
1                                   3   
2                                   3   
3                                   3   
4                                   3   

   Perovskite_composition_perovskite_ABC3_structure  \
0                                               1.0   
1                                               1.0   
2                                               1.0   
3                                               1.0   
4                                               1.0   

   Perovskite_composition_perovskite_inspired_structure  \
0                                                0.0      
1                                                0.0      
2                                                0.0      
3                                                0.0      
4                                                0.0      

  Perovskite_composition_a_ions  ...  \
0                            Cs  ...   
1                            Cs  ...   
2                            Cs  ...   
3                            Cs  ...   
4                            Cs  ...   

  Perovskite_deposition_solvent_annealing_timing  \
0                                            NaN   
1                                            NaN   
2                                            NaN   
3                                            NaN   
4                                            NaN   

  Perovskite_deposition_solvent_annealing_solvent_atmosphere  \
0                                            Unknown           
1                                            Unknown           
2                                            Unknown           
3                                            Unknown           
4                                            Unknown           

  Perovskite_deposition_solvent_annealing_time  \
0                                      Unknown   
1                                      Unknown   
2                                      Unknown   
3                                      Unknown   
4                                      Unknown   

  Perovskite_deposition_solvent_annealing_temperature  \
0                                            Unknown    
1                                            Unknown    
2                                            Unknown    
3                                            Unknown    
4                                            Unknown    

  Perovskite_deposition_after_treatment_of_formed_perovskite  \
0                                                NaN           
1                                                NaN           
2                                                NaN           
3                                 

In [8]:
# Export the Perovskite only DataFrame
output_file_path = 'perovskiteOnlyDF.xlsx'  
df_filtered.to_excel(output_file_path, index=False)

In [10]:
df_filtered['Perovskite_additives_compounds'].nunique()

1003

In [12]:
df_filtered['Perovskite_additives_concentrations'].nunique()

714

In [13]:
df_filtered['Perovskite_additives_compounds'].isna().sum()

28818

In [14]:
df_filtered['Perovskite_additives_concentrations'].isna().sum()

37995

In [15]:
column_to_clean = 'Perovskite_additives_concentrations'
df_conc = df_filtered.dropna(subset= [column_to_clean])

A Total of 5030 data points with 69 features.

In [16]:
df_conc

Perovskite_single_crystal  Perovskite_dimension_0D  \
5                            0.0                      0.0   
6                            0.0                      0.0   
7                            0.0                      0.0   
8                            0.0                      0.0   
9                            0.0                      0.0   
...                          ...                      ...   
43019                        0.0                      0.0   
43020                        0.0                      0.0   
43021                        0.0                      0.0   
43022                        0.0                      0.0   
43023                        NaN                      NaN   

       Perovskite_dimension_2D  Perovskite_dimension_2D3D_mixture  \
5                          0.0                                0.0   
6                          0.0                                0.0   
7                          0.0                                0.0   
8                          0.0                                0.0   
9                          0.0                                0.0   
...                        ...                                ...   
43019                      0.0                                0.0   
43020                      0.0                                0.0   
43021                      0.0                                0.0   
43022                      0.0                                0.0   
43023                      NaN                                NaN   

       Perovskite_dimension_3D  Perovskite_dimension_3D_with_2D_capping_layer  \
5                         True                                            0.0   
6                         True                                            0.0   
7                         True                                            0.0   
8                         True                                            0.0   
9                         True                                            0.0   
...                        ...                                            ...   
43019                     True                                            0.0   
43020                     True                                            0.0   
43021                     True                                            0.0   
43022                     True                                            0.0   
43023                     True                                            NaN   

      Perovskite_dimension_list_of_layers  \
5                                       3   
6                                       3   
7                                       3   
8                                       3   
9                                       3   
...                                   ...   
43019                                   3   
43020                                   3   
43021                                   3   
43022                                   3   
43023                                 NaN   

       Perovskite_composition_perovskite_ABC3_structure  \
5                                                   1.0   
6                                                   1.0   
7                                                   1.0   
8                                                   1.0   
9                                                   1.0   
...                                                 ...   
43019                                               1.0   
43020                                               1.0   
43021                                               1.0   
43022                                               1.0   
43023                                               NaN   

       Perovskite_composition_perovskite_inspired_structure  \
5                                                    0.0      
6                                                    0.0      
7                                

In [24]:
df_conc.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5030 entries, 5 to 43023
Data columns (total 69 columns):
 #   Column                                                          Non-Null Count  Dtype  
---  ------                                                          --------------  -----  
 0   Perovskite_single_crystal                                       4920 non-null   float64
 1   Perovskite_dimension_0D                                         4920 non-null   float64
 2   Perovskite_dimension_2D                                         4920 non-null   float64
 3   Perovskite_dimension_2D3D_mixture                               4920 non-null   float64
 4   Perovskite_dimension_3D                                         5030 non-null   bool   
 5   Perovskite_dimension_3D_with_2D_capping_layer                   4920 non-null   float64
 6   Perovskite_dimension_list_of_layers                             4920 non-null   object 
 7   Perovskite_composition_perovskite_ABC3_structure  

In [23]:
# Export the almost final minimized DataFrame
output_file_path = 'almostFinal.xlsx'  
df_filtered.to_excel(output_file_path, index=False)

In [27]:
df_conc['Perovskite_dimension_3D'].value_counts()

True     4834
False     196
Name: Perovskite_dimension_3D, dtype: int64

In [29]:
df_conc['Perovskite_band_gap_graded'].value_counts()

False    5014
True       16
Name: Perovskite_band_gap_graded, dtype: int64